### PR TITLE
Introduce fromEgressProxyRule

### PR DESCRIPTION
**Pull request by jschwinger233**
This simple PR introduces a new ip rule without installing it, mainly meant to take care of cilium downgrade for the new ip rule. Please see the commit messages for details.

This PR (along with its backport PRs) is the stage one of #31984 to address the remaining IPsec leak issues. Please see that issue for the the context.

Signed-off-by: Zhichuan Liang <gray.liang@isovalent.com>

### CODE REVIEW

File: pkg/proxy/routes.go [hunk 1]
```diff
@@ -100,10 +100,19 @@ func removeToProxyRoutesIPv6() error {
...
-		Mark:     linux_defaults.MagicMarkIsProxy,
+		Mark:     linux_defaults.MagicMarkIngress,
+		Mask:     linux_defaults.MagicMarkHostMask,
+		Table:    linux_defaults.RouteTableFromProxy,
+		Protocol: linux_defaults.RTProto,
+	}
+
+	//nolint:unused // Routing rule for traffic from egress proxy.
+	fromEgressProxyRule = route.Rule{
+		Priority: linux_defaults.RulePriorityFromProxyIngress,
```
Comment (rgo3) [inline pkg/proxy/routes.go]: Should this be the Ingress rule priority? From a quick glance I see we also have priority settings for egress rules

**Review by rgo3 — APPROVED** [review]

File: pkg/proxy/routes.go [hunk 2]
```diff
@@ -100,10 +100,19 @@ func removeToProxyRoutesIPv6() error {
...
+	// Routing rule for traffic from ingress proxy.
+	fromIngressProxyRule = route.Rule{
+		Priority: linux_defaults.RulePriorityFromProxy,
+		Mark:     linux_defaults.MagicMarkIngress,
+		Mask:     linux_defaults.MagicMarkHostMask,
+		Table:    linux_defaults.RouteTableFromProxy,
+		Protocol: linux_defaults.RTProto,
+	}
+
+	//nolint:unused // Routing rule for traffic from egress proxy.
```
Comment (julianwiedmann) [inline pkg/proxy/routes.go]:
```suggestion
	//nolint:unused // Routing rule for traffic from egress proxy.
```

huh, why is the `unused` needed? We *are* using it below, no?

File: pkg/proxy/proxy.go [hunk 3]
```diff
@@ -411,19 +411,22 @@ func (p *Proxy) ReinstallRoutingRules() error {
...
 				return err
 			}
 		} else {
 			if err := installFromProxyRoutesIPv4(node.GetInternalIPv4Router(), defaults.HostDevice); err != nil {
 				return err
 			}
 		}
+		if err := removeFromEgressProxyRoutesIPv4(); err != nil {
+			return err
+		}
```
Comment (julianwiedmann) [inline pkg/proxy/proxy.go]: This feels a bit awkward, as `removeFromProxyRoutesIPv4()` already deletes the whole `RouteTableFromProxy` table. 

Should we maybe put the call to `removeFromEgressProxyRoutesIPv4()` *prior* to the `if !option.Config.EnableIPSec || option.Config.TunnelingEnabled()` handling?

**Review by julianwiedmann — COMMENTED** [review]
ty! Sorry for taking so long - two inline comments, depends on your response whether they should be "request changes" :slightly_smiling_face:

File: pkg/proxy/routes.go (hunk 2 above)
Comment (jschwinger233) [reply pkg/proxy/routes.go]: Because it's in a separate commit. Should I make them together?

File: pkg/proxy/proxy.go (hunk 3 above)
Comment (jschwinger233) [reply pkg/proxy/proxy.go]: Sure!

File: pkg/proxy/routes.go (hunk 2 above)
Comment (julianwiedmann) [reply pkg/proxy/routes.go]: ah! No, then I think it's ok, but the last commit should remove the `nolint` again?

**Review by julianwiedmann — APPROVED** [review]
lgtm, thank you!

File: pkg/proxy/proxy.go [hunk 4]
```diff
@@ -410,6 +410,9 @@ func (p *Proxy) ReinstallRoutingRules() error {
 			return err
 		}
 
+		if err := removeFromEgressProxyRoutesIPv4(); err != nil {
```
Comment (tommyp1ckles) [inline pkg/proxy/proxy.go:413]: Does this mean that this change, and the subsequent one to add the egress proxy rules must be in introduced in different patch versions?

**Review by tommyp1ckles — APPROVED** [review]

File: pkg/proxy/proxy.go (hunk 4 above)
Comment (jschwinger233) [reply pkg/proxy/proxy.go:413]: Yes, will be backported to all stable branches.

**Pull request merged**